lib/Tapable.js

```diff
@@ -84,11 +84,13 @@ Tapable.prototype.applyPlugins2 = function applyPlugins2(name, param1, param2) {
 
 Tapable.prototype.applyPluginsWaterfall = function applyPluginsWaterfall(name, init) {
 	if(!this._plugins[name]) return init;
-	var args = Array.prototype.slice.call(arguments, 2);
+	var args = Array.prototype.slice.call(arguments, 1);
```

Comment:
Ok so as it stands we're cutting the arguments array down (with `slice`) to only the arguments that are shared between all plugin handler functions. Then, right before we actually pass these arguments in (with `apply` in the for loop) we quickly add the return value of the last function to these arguments (with `concat`). My logic was: to avoid doing concat over and over again thus constructing new arrays, we can just keep re-using the `args` array, while updating its first entry to be the return value of the previous handler `args[0] = current`. The end result is the same, but we get to keep re-using the same updatd array instead of calling `concat` over and over again when all `n-1` elements of the array are staying the same.

Does this sort of make sense? If not I'd be happy to try and re-explain ;)